src/Libraries/DSCPython/CPythonEvaluator.cs

```diff
@@ -151,10 +150,74 @@ public static class CPythonEvaluator
         private const string NodeName = "__dynamonodename__";
         static PyScope globalScope;
         internal static readonly string globalScopeName = "global";
+        internal static DynamoLogger DynamoLogger { get; set; }
 
         static CPythonEvaluator()
         {
             InitializeEncoders();
+            Dynamo.Models.DynamoModel.RequestPythonReset += RequestPythonRestartHandler;
+            // Session is null when running unit tests.
+            if (ExecutionEvents.ActiveSession != null)
+            {
+                DynamoLogger = ExecutionEvents.ActiveSession.GetParameterValue(ParameterKeys.Logger) as DynamoLogger;
+            }
+        }
+
+        internal static void RequestPythonRestartHandler(string pythonEngine)
+        {
+            //check if python engine request is for this engine, and engine is currently started
+            if (PythonEngine.IsInitialized && pythonEngine == "CPython3")
+            {
+                DynamoLogger?.Log("attempting reload of cpython3 modules", LogLevel.Console);
+                using (Py.GIL())
+                {
+                //the following is inspired by:
+                //TODO add attribution to aboutbox/license. clause3 BSD3 (preapproved)
+                //https://github.com/ipython/ipython/blob/master/IPython/extensions/autoreload.py
+                    var global = PyScopeManager.Global.Get(CPythonEvaluator.globalScopeName);
+                    global?.Exec(@"import sys
+import importlib
+import importlib.util
+import os
+def getInfoFile(module):
+    if not hasattr(module, '__file__') or module.__file__ is None:
```

Comment:
maybe add comments to all these filters/conditions so that we know what the intention was (in case something goes wrong)

Reply:
good idea, will add some comments

Reply:
where did you find documentation for these module attributes ?

Reply:
Maybe we should add a comment (like in the importlib.reload documentation)
```Reloading sys, __main__, builtins and other key modules is not recommended```
https://docs.python.org/3.4/library/importlib.html#importlib.reload

Reply:
will do - these attributes (and this function) were modified from the autoreload.py script which is why I added attribution to this library in the license and about boxes.